File: swift/Examples/PayKitDemo/PayKitDemo/SessionStream.swift
import Foundation
import SolanaPayKit

/// Drives a full MPP payment-channel **session** against the playground's
/// `/api/v1/stream` (a metered SSE endpoint), the flow the one-shot charge
/// client can't do:
///
///   1. GET the resource unauthenticated → 402 `WWW-Authenticate` session
///      challenge.
///   2. Open the channel: `PaymentChannelSession.open` builds the payer-signed
///      open transaction; the credential carries it and the server (operator)
///      co-signs + broadcasts. Retrying the GET with that credential returns the
///      200 `text/event-stream`.
///   3. Read the SSE deliveries (`data: {chunk,cost}` … `[DONE]`), summing cost.
///   4. Reserve a delivery on the side channel, then sign + commit one
///      cumulative voucher through `SessionConsumer`.
///   5. Optionally poll the receipt route for the on-chain settle signature.
///
/// Mirrors the TypeScript `SessionFetch` transport, scoped to what the demo
/// needs (aggregate the stream into a single voucher rather than throttled
/// live commits).
enum SessionStream {
    struct Result {
        let channelId: String
        let chunks: Int
        let totalPaidBaseUnits: UInt64
        let cumulative: String
        let settleSignature: String?
        /// Ordered trace of each step's output, surfaced in the log so the flow
        /// is visible (open -> stream -> commit -> settle).
        let steps: [String]
    }

    /// Run the session against `streamURL` (e.g. `<playground>/api/v1/stream`),
    /// paying with `payer` (the funded demo account). Returns a summary for the
    /// log. The voucher signer is a fresh ephemeral key (the on-chain authorized
    /// signer for this channel).
    static func consume(streamURL: URL, payer: SolanaSigner) async throws -> Result {
        let session = URLSession.shared
        var steps: [String] = []

        // 1. Unauthenticated GET → 402 session challenge.
        let challenge = try await fetchChallenge(streamURL, using: session)
        try challenge.requireSolanaSession()
        let request = try challenge.sessionRequest
        guard let blockhash = request.recentBlockhash, !blockhash.isEmpty else {
            throw SessionStreamError.message("session challenge did not carry a recentBlockhash")
        }
        guard let recentSlot = request.recentSlot else {
            throw SessionStreamError.message("session challenge did not carry a recentSlot")
        }

        // 2. Open the channel (pull + clientVoucher, server-broadcast).
        let sessionSigner = try MemorySigner(secretKey: randomSeed())
        let opener = try await PaymentChannelSession.open(
            request: request,
            payerSigner: payer,
            sessionSigner: sessionSigner,
            recentBlockhash: blockhash,
            openSlot: recentSlot
        )
        let channelId = opener.open.channelId.base58
        let credential = try serializeSessionCredential(challenge: challenge.echo(), action: opener.action)
        steps.append("opened channel \(shortId(channelId)) · deposit \(usd(request.cap))")

        // 3. Retry the GET with the open credential → 200 SSE; read deliveries.
        let (chunks, totalCost) = try await readStream(streamURL, authorization: credential, using: session)
        steps.append("streamed \(chunks) chunks · metered \(usd(totalCost))")

        // 4. Reserve one aggregate delivery, then sign + commit the voucher.
        var cumulative = "0"
        if totalCost > 0 {
            let originDeliveries = URL(string: "/__402/session/deliveries", relativeTo: streamURL)?.absoluteURL
                ?? streamURL.deletingLastPathComponent().appendingPathComponent("__402/session/deliveries")
            let commitURL = URL(string: "/__402/session/commit", relativeTo: streamURL)?.absoluteURL
                ?? originDeliveries

            let directive = try await reserveDelivery(
                originDeliveries,
                channelId: channelId,
                amount: totalCost,
                commitURL: commitURL.absoluteString,
                using: session
            )
            let consumer = SessionConsumer(
                session: opener.session,
                transport: HttpCommitTransport(commitURL: commitURL, urlSession: session)
            )
            let receipt = try await consumer.commitDirective(directive)
            cumulative = receipt.cumulative
            steps.append("committed voucher · cumulative \(receipt.cumulative) (\(receipt.status.rawValue))")
        }

        // 5. Best-effort receipt poll for the settle signature (server idle-closes).
        let settle = try? await pollReceipt(streamURL: streamURL, channelId: channelId, using: session)
        steps.append(settle != nil ? "settled on-chain · \(shortId(settle!))" : "settle pending (idle-close runs server-side)")

        return Result(
            channelId: channelId,
            chunks: chunks,
            totalPaidBaseUnits: totalCost,
            cumulative: cumulative,
            settleSignature: settle,
            steps: steps
        )
    }

    // MARK: - Steps

    private static func fetchChallenge(_ url: URL, using session: URLSession) async throws -> PaymentChallenge {
        var req = URLRequest(url: url)
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        let (_, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw SessionStreamError.message("no HTTP response from \(url)")
        }
        guard http.statusCode == 402 else {
            throw SessionStreamError.message("expected 402 from \(url.lastPathComponent), got \(http.statusCode)")
        }
        guard let header = http.value(forHTTPHeaderField: "Www-Authenticate") else {
            throw SessionStreamError.message("402 had no WWW-Authenticate header")
        }
        return try MppHeaders.parseWWWAuthenticate(header)
    }

    private static func readStream(
        _ url: URL,
        authorization: String,
        using session: URLSession
    ) async throws -> (chunks: Int, totalCost: UInt64) {
        var req = URLRequest(url: url)
        req.setValue(authorization, forHTTPHeaderField: "Authorization")
        req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 60

        let (bytes, response) = try await session.bytes(for: req)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw SessionStreamError.message("stream open failed: HTTP \(code)")
        }

        var chunks = 0
        var total: UInt64 = 0
        for try await line in bytes.lines {
            guard line.hasPrefix("data:") else { continue }
            let payload = line.dropFirst(5).trimmingCharacters(in: .whitespaces)
            if payload == "[DONE]" { break }
            guard let data = payload.data(using: .utf8),
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { continue }
            chunks += 1
            if let cost = obj["cost"] as? String, let value = UInt64(cost) {
                total += value
            }
        }
        return (chunks, total)
    }

    private static func reserveDelivery(
        _ url: URL,
        channelId: String,
        amount: UInt64,
        commitURL: String,
        using session: URLSession
    ) async throws -> MeteringDirective {
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        let body: [String: Any] = [
            "amount": String(amount),
            "sessionId": channelId,
            "deliveryId": "mpp-\(UUID().uuidString)",
            "commitUrl": commitURL,
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw SessionStreamError.message("delivery reservation failed: HTTP \(code)")
        }
        return try JSONDecoder().decode(MeteringDirective.self, from: data)
    }

    private static func pollReceipt(
        streamURL: URL,
        channelId: String,
        using session: URLSession
    ) async throws -> String? {
        guard let receiptURL = URL(string: "/sessions/receipt/\(channelId)", relativeTo: streamURL)?.absoluteURL else {
            return nil
        }
        // The server idle-closes after a short delay; poll a few times.
        for _ in 0..<8 {
            try await Task.sleep(nanoseconds: 1_500_000_000)
            let (data, response) = try await session.data(from: receiptURL)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200,
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { continue }
            // The program renamed finalize -> seal; accept either receipt flag
            // while playground servers roll over.
            let sealed = (obj["sealed"] as? Bool) ?? (obj["finalized"] as? Bool) ?? false
            if sealed, let sig = obj["settledSignature"] as? String, !sig.isEmpty {
                return sig
            }
        }
        return nil
    }

    private static func randomSeed() -> Data {
        Data((0..<32).map { _ in UInt8.random(in: 0...255) })
    }

    private static func shortId(_ value: String) -> String {
        value.count > 14 ? "\(value.prefix(6))…\(value.suffix(6))" : value
    }

    private static func usd(_ baseUnits: UInt64) -> String {
        let dollars = Decimal(baseUnits) / 1_000_000
        return "$\(NSDecimalNumber(decimal: dollars))"
    }

    private static func usd(_ baseUnitsString: String) -> String {
        UInt64(baseUnitsString).map(usd) ?? baseUnitsString
    }
}

/// Posts signed vouchers to the session commit side channel
/// (`POST /__402/session/commit`) and decodes the `CommitReceipt`.
private struct HttpCommitTransport: CommitTransport {
    let commitURL: URL
    let urlSession: URLSession

    func commit(directive: MeteringDirective, payload: CommitPayload) async throws -> CommitReceipt {
        var req = URLRequest(url: commitURL)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.httpBody = try JSONEncoder().encode(payload)
        let (data, response) = try await urlSession.data(for: req)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            let text = String(data: data, encoding: .utf8) ?? ""
            throw SessionStreamError.message("voucher commit failed: HTTP \(code) \(text)")
        }
        return try JSONDecoder().decode(CommitReceipt.self, from: data)
    }
}

enum SessionStreamError: Error, LocalizedError {
    case message(String)
    var errorDescription: String? {
        switch self {
        case .message(let text): return text
        }
    }
}

File: kotlin/examples/AndroidDemo/app/src/main/java/com/solana/paykit/demo/SessionStream.kt
package com.solana.paykit.demo

import com.solana.paykit.paycore.MemorySigner
import com.solana.paykit.paycore.MppException
import com.solana.paykit.paycore.SolanaSigner
import com.solana.paykit.protocols.mpp.client.CommitTransport
import com.solana.paykit.protocols.mpp.client.PaymentChannelSession
import com.solana.paykit.protocols.mpp.client.SessionConsumer
import com.solana.paykit.protocols.mpp.core.CommitPayload
import com.solana.paykit.protocols.mpp.core.CommitReceipt
import com.solana.paykit.protocols.mpp.core.MeteringDirective
import com.solana.paykit.protocols.mpp.core.MppHeaders
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.security.SecureRandom
import java.util.UUID

/**
 * Drives a full MPP payment-channel **session** against the playground's
 * `/api/v1/stream` (a metered SSE endpoint) — the flow the one-shot charge
 * client can't do. Mirrors the iOS `SessionStream`:
 *
 *   1. GET unauthenticated -> 402 `WWW-Authenticate` session challenge.
 *   2. Open the channel: `PaymentChannelSession.open` builds the payer-signed
 *      open tx; the credential carries it and the server (operator) co-signs +
 *      broadcasts. Retrying the GET with that credential returns the SSE stream.
 *   3. Read the SSE deliveries (`data: {chunk,cost}` ... `[DONE]`), summing cost.
 *   4. Reserve a delivery on the side channel, then sign + commit one cumulative
 *      voucher through `SessionConsumer`.
 *   5. Poll the receipt route for the on-chain settle signature.
 *
 * Synchronous (blocking OkHttp); call it off the main thread.
 */
object SessionStream {
    data class Result(
        val channelId: String,
        val chunks: Int,
        val totalPaidBaseUnits: Long,
        val cumulative: String,
        val settleSignature: String?,
        /** Ordered trace of each step's output, surfaced in the log so the flow
         *  is visible (open -> stream -> commit -> settle). */
        val steps: List<String>,
    )

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
        explicitNulls = false
    }
    private val JSON_MEDIA = "application/json".toMediaType()

    fun consume(client: OkHttpClient, streamUrl: String, payer: SolanaSigner): Result {
        val steps = mutableListOf<String>()
        // 1. Unauthenticated GET -> 402 session challenge.
        val challenge = client.newCall(
            Request.Builder().url(streamUrl).header("Accept", "application/json").get().build()
        ).execute().use { resp ->
            if (resp.code != 402) error("expected 402 from stream, got ${resp.code}")
            val header = resp.header("WWW-Authenticate") ?: error("402 had no WWW-Authenticate header")
            MppHeaders.parseWWWAuthenticate(header)
        }
        PaymentChannelSession.requireSolanaSession(challenge)
        val request = PaymentChannelSession.sessionRequest(challenge)
        val blockhash = request.recentBlockhash
        require(!blockhash.isNullOrEmpty()) { "session challenge did not carry a recentBlockhash" }
        // The server prefetches the current slot into the challenge as
        // `recentSlot`; it is the program's open_slot channel PDA seed the open
        // echoes back, only accepted within the 1500-slot open window.
        val recentSlot = request.recentSlot
        requireNotNull(recentSlot) { "session challenge did not carry a recentSlot" }

        // 2. Open the channel (pull + clientVoucher, server-broadcast).
        val sessionSigner = MemorySigner.fromSeed(randomSeed())
        val opener = PaymentChannelSession.open(request, payer, sessionSigner, blockhash, recentSlot)
        val channelId = opener.open.channelId.toBase58()
        val credential = PaymentChannelSession.serializeSessionCredential(challenge.echo(), opener.action)
        steps.add("opened channel ${shortId(channelId)} · deposit ${usd(request.cap)}")

        // 3. Retry the GET with the open credential -> 200 SSE; read deliveries.
        var chunks = 0
        var total = 0L
        client.newCall(
            Request.Builder().url(streamUrl)
                .header("Authorization", credential)
                .header("Accept", "text/event-stream")
                .get().build()
        ).execute().use { resp ->
            if (!resp.isSuccessful) error("stream open failed: HTTP ${resp.code}")
            val reader = resp.body?.charStream()?.buffered() ?: error("stream had no body")
            while (true) {
                val line = reader.readLine() ?: break
                if (!line.startsWith("data:")) continue
                val payload = line.removePrefix("data:").trim()
                if (payload == "[DONE]") break
                val obj = runCatching { json.parseToJsonElement(payload).jsonObject }.getOrNull() ?: continue
                chunks++
                (obj["cost"] as? JsonPrimitive)?.contentOrNull?.toLongOrNull()?.let { total += it }
            }
        }
        steps.add("streamed $chunks chunks · metered ${usd(total)}")

        // 4. Reserve one aggregate delivery, then sign + commit the voucher.
        var cumulative = "0"
        if (total > 0) {
            val directive = reserveDelivery(client, sideChannel(streamUrl, "/__402/session/deliveries"), channelId, total, streamUrl)
            val consumer = SessionConsumer(opener.session, HttpCommitTransport(client, sideChannel(streamUrl, "/__402/session/commit")))
            val receipt = consumer.commitDirective(directive)
            cumulative = receipt.cumulative
            steps.add("committed voucher · cumulative ${receipt.cumulative} (${receipt.status.name.lowercase()})")
        }

        // 5. Best-effort receipt poll for the settle signature.
        val settle = runCatching { pollReceipt(client, sideChannel(streamUrl, "/sessions/receipt/$channelId")) }.getOrNull()
        steps.add(if (settle != null) "settled on-chain · ${shortId(settle)}" else "settle pending (idle-close runs server-side)")

        return Result(channelId, chunks, total, cumulative, settle, steps)
    }

    private fun shortId(value: String): String =
        if (value.length > 14) "${value.take(6)}…${value.takeLast(6)}" else value

    private fun usd(baseUnits: Long): String =
        "$" + java.math.BigDecimal(baseUnits).divide(java.math.BigDecimal(1_000_000)).stripTrailingZeros().toPlainString()

    private fun usd(baseUnitsString: String): String =
        baseUnitsString.toLongOrNull()?.let { usd(it) } ?: baseUnitsString

    private fun reserveDelivery(
        client: OkHttpClient,
        url: String,
        channelId: String,
        amount: Long,
        commitUrl: String,
    ): MeteringDirective {
        val body = buildJsonObject {
            put("amount", amount.toString())
            put("sessionId", channelId)
            put("deliveryId", "mpp-${UUID.randomUUID()}")
            put("commitUrl", commitUrl)
        }
        val req = Request.Builder().url(url)
            .header("Accept", "application/json")
            .post(json.encodeToString(JsonObject.serializer(), body).toRequestBody(JSON_MEDIA))
            .build()
        return client.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) error("delivery reservation failed: HTTP ${resp.code}")
            json.decodeFromString(MeteringDirective.serializer(), resp.body!!.string())
        }
    }

    private fun pollReceipt(client: OkHttpClient, url: String): String? {
        repeat(8) {
            Thread.sleep(1_500)
            val obj = client.newCall(Request.Builder().url(url).get().build()).execute().use { resp ->
                if (!resp.isSuccessful) return@use null
                runCatching { json.parseToJsonElement(resp.body!!.string()).jsonObject }.getOrNull()
            } ?: return@repeat
            // The program renamed finalize -> seal; accept both receipt keys
            // while the playground still serves the legacy `finalized` flag.
            val sealed = (obj["sealed"] as? JsonPrimitive)?.contentOrNull == "true" ||
                (obj["finalized"] as? JsonPrimitive)?.contentOrNull == "true"
            val sig = (obj["settledSignature"] as? JsonPrimitive)?.contentOrNull
            if (sealed && !sig.isNullOrEmpty()) return sig
        }
        return null
    }

    private fun sideChannel(streamUrl: String, path: String): String =
        streamUrl.toHttpUrl().newBuilder().encodedPath(path).query(null).build().toString()

    private fun randomSeed(): ByteArray = ByteArray(32).also { SecureRandom().nextBytes(it) }
}

/** Posts signed vouchers to `POST /__402/session/commit` and decodes the receipt. */
private class HttpCommitTransport(
    private val client: OkHttpClient,
    private val commitUrl: String,
) : CommitTransport {
    private val json = Json { encodeDefaults = false; explicitNulls = false; ignoreUnknownKeys = true }
    private val media = "application/json".toMediaType()

    override fun commit(directive: MeteringDirective, payload: CommitPayload): CommitReceipt {
        val req = Request.Builder().url(commitUrl)
            .header("Accept", "application/json")
            .post(json.encodeToString(CommitPayload.serializer(), payload).toRequestBody(media))
            .build()
        return client.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) {
                throw MppException.InvalidTransaction(
                    "voucher commit failed: HTTP ${resp.code} ${resp.body?.string().orEmpty()}"
                )
            }
            json.decodeFromString(CommitReceipt.serializer(), resp.body!!.string())
        }
    }
}
